### PR TITLE
Classifying and Adding Simplifiers

### PR DESCRIPTION
**Pull request by saqzari**
Had to replicate termSize from Rankings.lhs into Classifier.lhs since importing causes an import loop

### CODE REVIEW

**Comment by andrewbutterfield** [comment]
What modules where involved in the import loop?

**Comment by saqzari** [comment]
> What modules where involved in the import loop?

```
  Module imports form a cycle:
  module `Classifier' (src\Classifier.lhs)
  imports `Ranking' (src\Ranking.lhs)
  which imports `LiveProofs' (src\LiveProofs.lhs)
  which imports `Theories' (src\Theories.lhs)
  which imports `Classifier' (src\Classifier.lhs)
```

**Comment by saqzari** [comment]
Anything else that should be done?

File: src/Classifier.lhs [hunk 1]
```diff
@@ -44,6 +46,33 @@ showAuto alaws = "   1. simps   -> "   ++ showSimps (simps alaws)  ++ "\n"
...
+
+addSimp :: String -> Term -> [(String, Direction)]
+addSimp nme (PCons sb n (p:q:xs)) = do
+                                    let sizeP = termSize p
+                                    let sizeQ = termSize q
+                                    if sizeP > sizeQ
+                                      then [(nme, Leftwards)] 
+                                    else if sizeP < sizeQ
+                                      then [(nme, Rightwards)]
+                                    else []
```
Comment (andrewbutterfield) [inline src/Classifier.lhs]: the lines 70-76 are just right - it's line 69 that needs to be changed and a conditional added to look  for the equivalence name

**Comment by andrewbutterfield** [comment]
Apologies - I had a deadline yeaterday so was a bit pre-occupied.

It looks to me that Ranking is probably importing stuff it shouldn't - but that's for me to fix.

File: src/Classifier.lhs [hunk 2]
```diff
@@ -44,6 +46,33 @@ showAuto alaws = "   1. simps   -> "   ++ showSimps (simps alaws)  ++ "\n"
...
+termSize (Lam _ _ vl t)       =  1 + length vl + termSize t
+termSize (Cls _ t)            =  1 + termSize t
+termSize (Sub _ t subs)       =  1 + termSize t + subsSize subs
+termSize (Iter _ _ _ _ _ vl)  =  3 + length vl
+termSize (Typ _)              =  2
+
+subsSize (Substn ts lvs)      =  3 * S.size ts + 2 * S.size lvs
+
+addSimp :: String -> Term -> [(String, Direction)]
+addSimp nme (PCons sb n (p:q:xs)) = do
```
Comment (andrewbutterfield) [inline src/Classifier.lhs]: I added a comment here that then got deleted!!!!

You need to pattern match for a list of length 2 only - and to chec that the name (`n`) matches that for equivalence.  The equivalence Identifier is obtained from the `theEqv` field of the `LogicSig` component of `REqState`.

Comment (saqzari) [reply src/Classifier.lhs:69]: Do i need to include logicSig in the method?

Comment (saqzari) [reply src/Classifier.lhs]: Alright, I have committed changes that should pattern match more specifically for list of length 2 and for equivalence

**Comment by andrewbutterfield** [comment]
This comment didn't occur in Github until I respodned to it via email???

In principle, you should compare `n` with `req.logicsig.theEqv`, but keep it like this for now.

--------------------------------------------------------------------
Andrew Butterfield     Tel: +353-1-896-2517     Fax: +353-1-677-2204
***@***.***, Head of Software Foundations & Verification Research Group
School of Computer Science and Statistics,
Room G.39, O'Reilly Institute, Trinity College, University of Dublin
                         http://www.scss.tcd.ie/Andrew.Butterfield/
--------------------------------------------------------------------


From: saqzari ***@***.***>
Reply to: andrewbutterfield/reasonEq ***@***.***>
Date: Saturday 4 March 2023 at 18:44
To: andrewbutterfield/reasonEq ***@***.***>
Cc: ***@***.***" ***@***.***>, Comment ***@***.***>
Subject: Re: [andrewbutterfield/reasonEq] Classifying and Adding Simplifiers (PR #6)


@saqzari commented on this pull request.

________________________________

In src/Classifier.lhs<https://github.com/andrewbutterfield/reasonEq/pull/6#discussion_r1125513599>:

> +

+termSize :: Term -> Int

+termSize (Val _ _)            =  1

+termSize (Var _ _)            =  1

+termSize (Cons _ _ _ ts)      =  1 + sum (map termSize ts)

+termSize (Bnd _ _ vs t)       =  1 + S.size vs + termSize t

+termSize (Lam _ _ vl t)       =  1 + length vl + termSize t

+termSize (Cls _ t)            =  1 + termSize t

+termSize (Sub _ t subs)       =  1 + termSize t + subsSize subs

+termSize (Iter _ _ _ _ _ vl)  =  3 + length vl

+termSize (Typ _)              =  2

+

+subsSize (Substn ts lvs)      =  3 * S.size ts + 2 * S.size lvs

+

+addSimp :: String -> Term -> [(String, Direction)]

+addSimp nme (PCons sb n (p:q:xs)) = do

Do i need to include logicSig in the method?

**Comment by andrewbutterfield** [comment]
If I start a proof and invoke `au` I don't see any output. You need to apply this to all the laws(*) and output the results.

(*) first just the laws in the current theory - the one where the proof resides. Then aplly alto to all laws in theories that are imported by this one.

**Comment by saqzari** [comment]
> If I start a proof and invoke `au` I don't see any output. You need to apply this to all the laws(*) and output the results.
> 
> (*) first just the laws in the current theory - the one where the proof resides. Then aplly alto to all laws in theories that are imported by this one.

Could you go into more detail about this? I currently have the classify command which puts the inputted law into the simps. Should au invoke one of them?

**Comment by andrewbutterfield** [comment]
Apologies – I got confused.

I did  run classify – it works fine, however…
When you run the command, it should print out what classification it gave the law, if any.
I had to do a ‘show T’ afterwards to see changes.

You should now look at implementing a simplification process.

Some nit-picking


`
AutoLaws:

   1. simps   -> (false_def,Rightwards)(lnot_invol,Leftwards)(false_neg,Leftwards)(lnot_def,Rightwards)(false_subst,Leftwards)

   2. folds   ->

   3. unfolds ->
`

Can you do a linebreak after each entry, and indent nicely?

Can you add the option ‘.’ To classify which will classify all laws in the current theory,
And ‘*’ to classify all laws in dependent theories
(Look at Assume/Demote commands)




From: saqzari ***@***.***>
Reply to: andrewbutterfield/reasonEq ***@***.***>
Date: Monday 6 March 2023 at 15:19
To: andrewbutterfield/reasonEq ***@***.***>
Cc: ***@***.***" ***@***.***>, Comment ***@***.***>
Subject: Re: [andrewbutterfield/reasonEq] Classifying and Adding Simplifiers (PR #6)


If I start a proof and invoke au I don't see any output. You need to apply this to all the laws(*) and output the results.

(*) first just the laws in the current theory - the one where the proof resides. Then aplly alto to all laws in theories that are imported by this one.

Could you go into more detail about this? I currently have the classify command which puts the inputted law into the simps. Should au invoke one of them?

**Comment by saqzari** [comment]
Hello, I have classify * and . integrated and also changed up sh T for simplifies. Let me know if there are issues

File: src/Theories.lhs [hunk 3]
```diff
@@ -499,15 +499,8 @@ upgrade cjnm thry sjc (cj@(nm,asn):cjs)
...
-lawClassify lw thry = return Theory {  thName   =  thName thry
-                                     , thDeps   =  thDeps thry
-                                     , known    =  known thry
-                                     , laws     =  laws thry
-                                     , proofs   =  proofs thry
-                                     , conjs    =  conjs thry
-                                     , auto     =  addLawsClassifier (lawNamedAssn lw) (auto thry)
-                                    }
+lawClassify :: MonadFail m => [Law] -> Theory -> m Theory
+lawClassify lw thry = return $ auto_ (addLawsClass (lw) (auto thry)) thry
```
Comment (andrewbutterfield) [inline src/Theories.lhs:503]: I think `return $ auto__ (addLawsClass lw) thry` should also work?

File: src/AbstractUI.lhs [hunk 4]
```diff
@@ -275,21 +275,26 @@ demoteLaw thnm whichL reqs
...
+ | lnm == "."      =  return lwsCur
  | null law1       =  fail ("law '"++lnm++"': not found")
  | otherwise       =  return [theLaw]
  where
-   lws = laws thry
-   (_,law1,_) = extract (((lnm==) . fst) .fst) lws
+   currTh = currTheory reqs
+   thys = theories reqs
+   lwsCur = laws thry
+   lwsDep = concat $ map laws (fromJust $ getTheoryDeps currTh thys)
```
Comment (andrewbutterfield) [inline src/AbstractUI.lhs:288]: This won't work - it adds the classification into the theory where the command was issued. 
It needs to process each dependent theory indivudually - see `assumeDepConj` in `Theories.lhs`.

Comment (andrewbutterfield) [reply src/AbstractUI.lhs:288]: I'm going to accept this anyway - the above isn't a deal breaker.

You need to focus now on applying these simplifiers in a proof.

**Pull request merged**